pyiceberg/io/pyarrow.py

```diff
@@ -1320,7 +1321,16 @@ def _cast_if_needed(self, field: NestedField, values: pa.Array) -> pa.Array:
                     and pa.types.is_timestamp(values.type)
                     and values.type.unit == "ns"
                 ):
-                    return values.cast(target_type, safe=False)
+                    if target_type.tz == "UTC" and values.type.tz in UTC_ALIASES or not target_type.tz and not values.type.tz:
+                        return values.cast(target_type, safe=False)
+                if (
+                    pa.types.is_timestamp(target_type)
+                    and target_type.unit == "us"
+                    and pa.types.is_timestamp(values.type)
+                    and values.type.unit in {"s", "ms", "us"}
```

Comment:
When we're requesting `us`and the file provide a `us`, I don't think we need to cast?
```suggestion
                    and values.type.unit in {"s", "ms"}
```

Reply:
Sorry this is a bit confusing, I agree 😓 . I included it here because I wanted to support casting `pa.timestamp('us', tz='Etc/UTC')` to `pa.timestamp('us', tz='UTC')` within the same condition.

I think we won't hit this condition if both the input and requested types are `pa.timestamp('us')` because we enter this block only if target_type and values.type are not equal:
https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1315

Reply:
I think we might need to do some more work here. In Iceberg we're rather strict on the distinction between Timestamp and TimestampTZ. A good way of showing this can be found here:

https://github.com/apache/iceberg-python/blob/aceed2ad9898eaa8b0cdddd883a195e58af5d0e3/pyiceberg/expressions/literals.py#L566-L572

This is when we parse a string from a literal, which often comes an expression: `dt >= '1925-05-22T00:00:00'`. If the value has a timezone, even UTC, we reject it as `Timestamp`. When it has a timestamp, we [normalize it to UTC and then store the integer](https://github.com/apache/iceberg-python/blob/aceed2ad9898eaa8b0cdddd883a195e58af5d0e3/pyiceberg/utils/datetime.py#L75).

Reply:
Sounds good @Fokko- thank you for the review.
I've made these checks stricter and also clearer for users to follow.